src/lexer.cc

```diff
@@ -0,0 +1,139 @@
+/*
+   Copyright 2019 Simon (psyomn) Symeonidis
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+ */
+#include "bk/lexer.h"
+#include "bk/token.h"
+
+#include <filesystem>
+#include <sstream>
+#include <fstream>
+
+namespace bk {
+static bool isPunct(char c) {
+  return
+    c == '{' || c == '}' ||
+    c == '.' || c == ',' ||
+    c == ';' || c == ':' ||
+    c == '\'' || c == '"' ||
+    c == '!' || c == '?' ||
+    c == '-' ||
+    c == '(' || c == ')' ||
+    c == '[' || c == ']' ||
+    c == '<' || c == '>';
+}
+
+static bool isSymbol(char c) {
+  return
+    c == '@' || c == '#' || c == '$' || c == '^' ||
+    c == '&' || c == '*' || c == '|' || c == '\\' ||
+    c == '/';
+}
+
+static bool isNewline(char c) {
+  return c == '\n';
+}
+
+static bool isWhitespace(char c) {
+  // ignore chars
+  return c == ' ' || c == '\t' || c == '\r';
+}
+
+Lexer::Lexer(const std::string& path) :
+  mLineNum(1),
+  mAcceptExtension(".md") {
+  mRoot = path;
+}
+
+Lexer::~Lexer() {}
+
+void Lexer::process() {
+  namespace fs = std::filesystem;
+
+  for (auto& el : fs::recursive_directory_iterator(mRoot))
+    if (el.path().extension() == mAcceptExtension) mFiles.push_back(el.path());
+
+  for (auto& path : mFiles) tokenizeFile(path);
+}
+
+void Lexer::tokenizeFile(const std::string& filename) {
+  std::ifstream file(filename, std::ifstream::in);
+  std::stringstream ss;
+  std::vector<bk::Token> tokens;
+  std::size_t linecount = 0;
+  char c;
+
+  while (file.get(c)) {
+    if (isWhitespace(c)) continue;
+
+    if (isNewline(c)) {
+      linecount++;
+      bk::Token tok(std::string(""), linecount);
+      tok.type(Token::Type::newline);
+      tokens.push_back(tok);
+
+      ss.str(std::string());
+      continue;
+    }
+
+    if (isPunct(c)) {
+      // punct = {},.;:'"!?-()[]<>
+      ss << c;
+      bk::Token tok(ss.str(), linecount);
```

Comment:
Note that you're invoking an extra copy on the `ss` string here. The reason being that `Token` accepts an lvalue reference which forces a copy but since the value here is an rvalue, you could use move semantics to skip the copy.

Performance impact here is minimal, but just pointing it out as this is the type of things you're constantly thinking about in C++. Note that if you want an intro to rvalue references and move semantics, you can find it in the more recent Meyers book, or I can give you an intro.

Reply:
This will be definitely something I need to look more into. Thank you!